### PR TITLE
[release-0.18] update for Julia v0.7

### PR DESCRIPTION
**Pull request by joehuchette**
This drops support for v0.6 for the moment

### CODE REVIEW

File: REQUIRE [hunk 1]
```diff
@@ -1,5 +1,8 @@
-julia 0.6
+julia 0.7
 MathProgBase 0.6 0.8
 ReverseDiffSparse 0.8 0.9
 ForwardDiff 0.5 0.8
 Calculus
+LinearAlgebra
```
Comment (mlubin) [inline REQUIRE]: I don't think these are needed, they're baked into Julia.

**Comment by mlubin** [comment]
You can just put `Compat` in front of the packages to make this work on 0.6.

File: REQUIRE (hunk 1 above)
Comment (joehuchette) [reply REQUIRE]: I think you're right

Comment (blegat) [reply REQUIRE]: You need to do `using Compat.LinearAlgebra` for it to work on both v0.6 and v0.7

File: src/JuMPContainer.jl [hunk 2]
```diff
@@ -211,9 +211,9 @@ mutable struct KeyIterator{JA<:JuMPArray}
     x::JA
     dim::Int
     next_k_cache::Array{Any,1}
-    function (::Type{KeyIterator{JA}}){JA}(d)
+    function (::Type{KeyIterator{JA}})(d) where JA
```
Comment (blegat) [inline src/JuMPContainer.jl]: It should just be `function KeyIterator{JA}(d) where JA` if I'm not mistaken

**Comment by joehuchette** [comment]
I believe it's working on both v0.6 and v0.7 now

File: src/solvers.jl [hunk 3]
```diff
@@ -250,6 +250,8 @@ function solve(m::Model; suppress_warnings=false,
             # Do a separate try since getobjval could work while getobjbound does not and vice versa
             objBound = MathProgBase.getobjbound(m.internalModel) + m.obj.aff.constant
             m.objBound = objBound
+        catch
+            nothing
```
Comment (mlubin) [inline src/solvers.jl]: Why is this needed?

Comment (joehuchette) [reply src/solvers.jl]: On v0.7 each try block needs either a ``catch`` or a ``finally`` condition

**Comment by mlubin** [comment]
Could you switch nightly to 0.7 on travis/appveyor and remove it as an allowed failure? There are still some errors.

File: appveyor.yml [hunk 4]
```diff
@@ -2,6 +2,8 @@ environment:
   matrix:
   - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.6/julia-0.6-latest-win32.exe"
   - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x64/0.6/julia-0.6-latest-win64.exe"
+  - JULIA_URL: "https://julialang-s3.julialang.org/bin/winnt/x86/0.7/julia-0.7.0-beta-win32.exe"
```
Comment (mlubin) [inline appveyor.yml]: Is there a `0.7-latest`?

**Comment by mlubin** [comment]
I don't see any regressions on 0.6, so we could merge and continue 0.7 work in separate PRs.

**Comment by joehuchette** [comment]
Feel free to merge if you don't want updating to block on this PR; I won't get the chance to return to this for a couple days.

**Comment by ufechner7** [comment]
Would be nice if this could be merged even if it not the final fix for 0.7 yet. I think we need an iterative approach to make the packages work.

**Comment by mlubin** [comment]
I agree, but I won't merge until I get a green light from travis on 0.6. This is blocked by (at least) https://github.com/JuliaLang/METADATA.jl/pull/15933.

**Comment by IainNZ** [comment]
METADATA PR in, can re-run CI?

**Comment by odow** [comment]
Not sure how to restart Appveyor? Do I need some extra permissions or something?

**Comment by blegat** [comment]
@odow I have restarted 30 minutes ago
